Add K8s and Calico IPv6 config

tmjd opened the pull request:
## Description

Adding docs to outline configuration changes needed to setup Kubernetes and Calico with IPv6.


## Todos

## Release Note

```release-note
None required
```

caseydavenport commented:
CC @valentin2105

valentin2105 commented:
Thanks for adding me, 
Great documentation with most of the step to use Calico/k8s in IPv6 !

I think we need to explain than ClusterIP will not work in IPv4 (for sure) so, the only solution I found is use `ClusterIP: None` on all services to create DNS names that point on `Endpoints` directly.

I haven't try with Calico v3 but it seem that an IPv6-only Node requiere for now an IPv4 (only for the Calico "registration" of the node).  
I will try today to add a Calico v3 Node without any IPv4 but I'm not sure it will work. 
If doesn't, I will take a look to jow-to fix this issue.

tmjd commented:
> I think we need to explain than ClusterIP will not work in IPv4 (for sure) so, the only solution I found is use ClusterIP: None on all services to create DNS names that point on Endpoints directly.

Is that only if someone is setting up an IPv4 K8s cluster but having Calico assign IPv6 addresses?  The recommendation put forward here is to have both K8s and Calico be using the same IP version.

> it seem than an IPv6-only Node requiere for now an IPv4

That could be the case, the systems I was testing on were IPv4 hosts that I added IPv6 addresses to.  I'm interested to hear about your results.

Thanks for taking a look @valentin2105 and for your feedback.

valentin2105 commented:
> Is that only if someone is setting up an IPv4 K8s cluster but having Calico assign IPv6 addresses? The recommendation put forward here is to have both K8s and Calico be using the same IP version.
Humm, my k8s cluster IP Setup in IPv6 but I'm don't think that k8s can manage IPv6 ClusterIP and manage their IP6Tables rules with Kube-Proxy.
Maybe I'm wrong, but the only kind of services that we can use in v6 is `ClusterIP: None` ?

Have your ever seen IPv6 ClusterIP and Kubeproxy that do the IPv6 job on IPTables ? 

> Thanks for taking a look @valentin2105 and for your feedback.

Really happy ;)

valentin2105 commented:
@tmjd 

Here is my test for add an IPv6-only node with Calico v3.0.3 on k8s v.1.9.3 : 

```
2018-03-07 23:38:36.130 [INFO][9] startup.go 248: Early log level set to info
2018-03-07 23:38:36.131 [INFO][9] startup.go 259: NODENAME environment not specified - check HOSTNAME
2018-03-07 23:38:36.202 [INFO][9] startup.go 101: Skipping datastore connection test
2018-03-07 23:38:36.208 [INFO][9] startup.go 332: Building new node resource Name="bruce.domain.tld"
2018-03-07 23:38:36.208 [INFO][9] startup.go 347: Initialize BGP data
2018-03-07 23:38:36.209 [WARNING][9] startup.go 540: Unable to auto-detect an IPv4 address: no valid IPv4 addresses found on the host interfaces
2018-03-07 23:38:36.209 [WARNING][9] startup.go 369: Couldn't autodetect an IPv4 address. If auto-detecting, choose a different autodetection method. Otherwise provide an explicit address.
2018-03-07 23:38:36.209 [WARNING][9] startup.go 1007: Terminating
Calico node failed to start
```
It's quite blocking.

tmjd commented:
Thanks for checking on the IPv6, do you know if there is a Github Issue for that? I was thinking there might be but can't find one for it.

I do see cluster IPs created by Kubernetes and iptables setting up rules for them.  I've included the output showing that in case there is something I'm misunderstanding.
I'm using Kubernetes 1.9.3 also.

```
core@k8s-master ~ $ kubectl get services --all-namespaces
NAMESPACE     NAME         TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)         AGE
default       kubernetes   ClusterIP   fd30::1      <none>        443/TCP         8m
kube-system   kube-dns     ClusterIP   fd30::10     <none>        53/UDP,53/TCP   1m
policy-demo   nginx        NodePort    fd30::6ec8   <none>        80:32184/TCP    1m
```

```
-A KUBE-SERVICES ! -s fd20::/112 -d fd30::6ec8/128 -p tcp -m comment --comment "policy-demo/nginx: cluster IP" -m tcp --dport 80 -j KUBE-MARK-MASQ
-A KUBE-SERVICES -d fd30::6ec8/128 -p tcp -m comment --comment "policy-demo/nginx: cluster IP" -m tcp --dport 80 -j KUBE-SVC-3VXIGVIYYFN7DHDA
-A KUBE-SERVICES ! -s fd20::/112 -d fd30::1/128 -p tcp -m comment --comment "default/kubernetes:https cluster IP" -m tcp --dport 443 -j KUBE-MARK-MASQ
-A KUBE-SERVICES -d fd30::1/128 -p tcp -m comment --comment "default/kubernetes:https cluster IP" -m tcp --dport 443 -j KUBE-SVC-NPX46M4PTMTKRN6Y
-A KUBE-SERVICES ! -s fd20::/112 -d fd30::10/128 -p udp -m comment --comment "kube-system/kube-dns:dns cluster IP" -m udp --dport 53 -j KUBE-MARK-MASQ
-A KUBE-SERVICES -d fd30::10/128 -p udp -m comment --comment "kube-system/kube-dns:dns cluster IP" -m udp --dport 53 -j KUBE-SVC-TCOU7JCQXEZGVUNU
-A KUBE-SERVICES ! -s fd20::/112 -d fd30::10/128 -p tcp -m comment --comment "kube-system/kube-dns:dns-tcp cluster IP" -m tcp --dport 53 -j KUBE-MARK-MASQ
-A KUBE-SERVICES -d fd30::10/128 -p tcp -m comment --comment "kube-system/kube-dns:dns-tcp cluster IP" -m tcp --dport 53 -j KUBE-SVC-ERIFXISQEP7F7OF4
```

tmjd commented:
Found the issue #1204

caseydavenport commented on master/usage/ipv6.md:
```diff
@@ -1,20 +1,155 @@
-security configuration, a VM can initiate an IPv6 connection to another
-VM, or to an IPv6 destination outside the data center; and that a VM can
-terminate an IPv6 connection from outside.
+{{site.prodname}} supports connectivity over IPv6, for hosts and
```
https://docs.projectcalico.org/v3.0/usage/decommissioning-a-node

I'd like to start aligning the structure of our usage documentation. ^ I think the decommissioning a node document is a good reference point to start with.

One important thing is that each document starts by concisely describing exactly why one might want to follow this document, so users can tell really quickly whether or not it is worthwhile for them to read the rest of the document.

caseydavenport commented on master/usage/ipv6.md:
```diff
@@ -1,20 +1,155 @@
+
+## Enabling IPv6 with Kubernetes
+
+Currently Kubernetes supports only one IP stack version at a time. This means
```
I think we should try to organize this sort of information into either a "Prerequisites" or "Limitations" section towards the top of the page, for reasons similar to what I mention in my comment above.

In doing so, it would be great if we could break it down from large paragraphs of text into shorter bullet points - that makes it a lot easier for people to parse.

caseydavenport commented on master/usage/ipv6.md:
```diff
@@ -1,20 +1,155 @@
+
+### Kubernetes Component Setup
+
+To deploy {{site.prodname}} with IPv6 on Kubernetes it is first necessary
```
This is probably also "prerequisites" section content.

caseydavenport commented on master/usage/ipv6.md:
```diff
@@ -1,20 +1,155 @@
+### DNS changes
+
+It will probably be necessary to modify your DNS pod for IPv6. If you are using
+[KubeDNS](/{{page.version}}/getting-started/kubernetes/installation/manifests/kubedns.yaml), 
```
KubeDNS -> `kube-dns`

caseydavenport commented on master/usage/ipv6.md:
```diff
@@ -1,20 +1,155 @@
+[KubeDNS](/{{page.version}}/getting-started/kubernetes/installation/manifests/kubedns.yaml), 
+then the following changes will ensure IPv6 operation.
+
+- Update the image versions to at least `1.14.8`.
```
It would be great if there was upstream Kubernetes documentation we could point to for all of this.... I'm guessing nothing exists for now but when it does we can clean this up a lot.

tmjd commented:
I've added a Note that IPv4 is required also and a link to the github bug that covers that.

I also rebased my changes off master so that should fix the test failure.

caseydavenport commented on master/usage/ipv6.md:
```diff
@@ -1,20 +1,158 @@
+> operation.
+{: .alert .alert-info}
+
+### Calico Manifest changes
```
I'd suggest making this more "verb-y" with a set of steps users can follow.

i.e. 

```
### Enabling IPv6 in Calico

1. Update IPAM to allocate IPv6 addresses

< Example here >

2. Configure Calico with the desired IPv6 IP Pool

< Example here >

3. Enable IPv6 data plane support

 < Example here >

```

tmjd replied:
I thought when we chatted we said a list would be more appropriate because it is just setting these values. If we're looking for a procedure style what about something like:
````
1. Edit your manifest making the following changes.
    <Bullet list>
2. Apply your manifest as you normally would.
```

I'm also not sure about the `<Example here>` pieces as these are changes needed in one large yaml which makes it tricky to take a snippet from.

valentin2105 commented:
@tmjd in your example, Is this cluster IPs provided by k8s ? with ` --service-cluster-ip-range=fd::/8` ?  
And the IP6tables rules are created by Kube-Proxy ? 

I need to try that.  😈  
I'm tired of changing all my services to `ClusterIP: None`.

tmjd commented:
@valentin2105 Yes I provided `--service-cluster-ip-range=fd30::0/110` to my kube-apiserver.

caseydavenport commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+
+### Limitations
+
+- Currently Kubernetes supports only one IP stack version at a time. This
```
nit: "assign only IPv6 address"

caseydavenport commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+- Currently Kubernetes supports only one IP stack version at a time. This
+  means that if you configure Kubernetes for IPv6 then {{site.prodname}}
+  should be configured to only assign IPv6 addresses.
+- You should not switch a deployment that is using IPv4 to IPv6, or
```
This sentence reads a bit awkwardly - perhaps "These steps have not been tested against an existing IPv4 cluster and are intended only for new clusters" or something like that

caseydavenport commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+Support. This would be because:
+- You want workloads to communicate over IPv6.
+- You want workloads to initiate connections to IPv6 services.
+- You want workloads to terminate incomint IPv6 connections.
```
typo: incomint

caseydavenport commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+### About IPv6 Support
+
+You may be interested in this document if you want to enable IPv6
+Support. This would be because:
```
Support -> support

caseydavenport replied:
"This would be because" -> "For example:"

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
 ---
-title: IPv6 Support
+title: Enabling IPv6 Support
```
Need to also update the side nav title to match

caseydavenport commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+
+### Details of Enabling IPv6
+
+To enable IPv6 operation of a {{site.prodname}} deployment follow the
```
"To enable IPv6 support when installing Calico" might be a simpler way to say this

caseydavenport commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+
+1. Download the Calico manifest with the calico-config ConfigMap and
+   calico-node Daemonset as `calico.yaml`.
+1. Update the `calico.yaml` file with the following changes:
```
Instead of a nested list here, I'd just make each of these a step - I think that will be easier to follow.

caseydavenport commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+       kube-controller-manager and to kube-proxy.
+   - Ensure the environment variable `FELIX_IPV6SUPPORT` is set `true`
+     on the calico-node Daemonset.
+1. Apply the `calico.yaml` manifest.
```
Maybe an example kubectl command here?

caseydavenport commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+- You should not switch a deployment that is using IPv4 to IPv6, or
+  vise-versa but to instead setup a new cluster.
+
+### Details of Enabling IPv6
```
I'd name this section "Enabling IPv6 support in Calico" or something to it's clear we're talking about required changes to calico here.

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+
+The sections below are not intended to be a detailed setup guide but
+to layout the settings necessary to enable IPv6 operation of Kubernetes
+and Calico.
```
This is a hail mary: lines 21, 22, 23....highly recommend deleting.

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+### About IPv6 Support
+
+You may be interested in this document if you want to enable IPv6
+Support. This would be because:
```
Please consider consolidating `### About IPv6 Support` and `### Purpose of this page`. This could be compressed into a single introduction section with links to get people directly to the section that they care about as follows:

```
## About enabling IPv6

After enabling IPv6:
- Workloads can communicate over IPv6.
- Workloads can initiate connections to IPv6 services.
- Workloads can terminate incoming IPv6 connections.

Support for IPv6 and the procedure for enabling it varies by orchestrator. Refer to the section that corresponds to your orchestrator for details.

- [Enabling IPv6 with Kubernetes](#enabling-ipv6-with-kubernetes)
- [Enabling IPv6 with OpenStack](#enabling-ipv6-with-openstack)
```

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+- Each host must have an IPv6 address that is reachable from the other
+  hosts.
+  > **Note:** Each host running Calico currently must also have an IPv4
+  > address due to [bug 1204](https://github.com/projectcalico/calico/issues/1204).
```
In general, it's better not to include known issues in the documentation. Better to confine the known issues in the Release Notes.  However, it is something that perhaps warrants mention. I guess if we do choose to include this here, which might be necessary, you should just add to the GitHub issue that when it is fixed, we need to manually update this note.

caseydavenport replied:
I wouldn't necessarily classify this as a "known issue" - it's just a "how it works" type thing which we probably will improve later.

I'd rephrase as a requirement "Each host must have an IPv6 address reachable from all other hosts as well as an IPv4 address"

tmjd replied:
@caseydavenport WDYT?

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+
+### Prerequisites
+
+#### Host Prerequisites
```
We are trying to standardize on sentence case for the headings, so "Prerequisites" should be lowercased -> "prerequisites"

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+#### Host Prerequisites
+
+- Each host must have an IPv6 address that is reachable from the other
+  hosts.
```
Is it all other hosts? Might be nice if you could be more specific here....

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+  and Calico traffic is forwarded appropriately.
+- Each host must have a default IPv6 route.
+
+#### Kubernetes Components Prerequisites
```
^^ Sentence case

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+> **Note:** These flags are not the only flags that will need to be set on the
+> kubernetes components but the ones that were found necessary to ensure IPv6
+> operation.
+{: .alert .alert-info}
```
I feel like this note (lines 44-47) does not really say anything that's not already in lines 41-43. Would recommend deleting lines 44-47.

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+#### Kubernetes Components Prerequisites
+
+Kubernetes components must be configured to operate with IPv6.
+Below are the flags for each Kubernetes component that will help ensure
```
why does it say "help"? Makes it sound like you're not giving them all the information they would need to enable ipv6...which is the title and purpose of the page....

what you could do is to rephrase lines 42-43 to this "To enable IPv6, set the following flags." 

That pretty much covers everything with a lot less words IMO

tmjd replied:
It is because I am not confident that the flags laid out are the only ones necessary for Kubernetes to use IPv6.  I was trying to give the impression that this is to help them in setting up K8s with IPv6 and not a comprehensive guide.

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+  - The `--master` flag should be set with the IPv6 address where the
+    kube-apiserver can be accessed.
+  - The `--cluster-cidr` flag must be set to match the {{site.prodname}}
+    IPv6 IPPool.
```
I don't want to complicate your efforts, but it feels like a tabular treatment would be better for the above. Just a minor point.

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+  - The `--cluster-cidr` flag must be set to match the {{site.prodname}}
+    IPv6 IPPool.
+
+### Limitations
```
I would suggest you bump this section up to the very beginning. It is too late for them when they get all the way down here

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+  means that if you configure Kubernetes for IPv6 then {{site.prodname}}
+  should be configured to only assign IPv6 addresses.
+- You should not switch a deployment that is using IPv4 to IPv6, or
+  vise-versa but to instead setup a new cluster.
```
"You should not switch a deployment that is using IPv4 to IPv6, or vise-versa but to instead setup a new cluster." --> "Do not switch a deployment using IPv4 to IPv6. Only enable IPv6 on a new cluster."

(btw, this should be almost the first sentence in the section)

tmjd replied:
Did you see Casey's comment about changing this sentence? https://github.com/projectcalico/calico/pull/1736#discussion_r173524437
I like it since it is more positive about what this page is conveying.  Also more accurate about what we have tested ourselves since I did not actually try to take a deployment and switch it from IPv4 to IPv6.

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+To enable IPv6 operation of a {{site.prodname}} deployment follow the
+steps below.
+
+1. Download the Calico manifest with the calico-config ConfigMap and
```
"Calico" --> "{{site.prodname}}"

tmjd replied:
There are multiple manifests in our docs that this applies to. Should I try to link to them all, I think there are at least 3 it would work with and possibly 2 more the canal ones that I'm not sure if it would.

caseydavenport replied:
I don't think we should link to a specific manifest, or all of them - there are just too many right now. We could link to the self-hosted install landing page maybe, or if we really need to I'd say use the kubeadm one.

tmjd replied:
My intention behind describing the manifest briefly was to try to make sure the proper manifest is being modified.  That may be overkill because I would hope someone reading this would realize they had the wrong manifest (like an RBAC one) when they couldn't find any of the places they needed to modify.

What if I change it to
`Download the {{site.prodname}} manifest you wish to update for IPv6 deployment and save it as 'calico.yaml'.`?

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+
+     - Set the `CALICO_IPV6POOL_CIDR` environment variable to the desired pool,
+       it should match the `--cluster-cidr` passed to the
+       kube-controller-manager and to kube-proxy.
```
I do not know if you intended to add an additional level of nesting to this list with the indentation of line 113? I imagine this was just an error. If not, I would recommend against adding this extra level of nesting to the list.

tmjd replied:
Yes it is an extra level of nesting because it applies to fields the above block is adding.

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+- You should not switch a deployment that is using IPv4 to IPv6, or
+  vise-versa but to instead setup a new cluster.
+
+### Details of Enabling IPv6
```
I would recommend changing this title to `### Enabling IPv6` and changing everything that follows to just a simple numbered procedure with a new number for each action they need to take and no nesting.

tmjd replied:
What do you think about Casey's suggestion of `Enabling IPv6 support in Calico`?

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+     on the calico-node Daemonset.
+1. Apply the `calico.yaml` manifest.
+
+### DNS changes
```
You can change this heading also to "Modifying your DNS for IPv6"

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+
+### Prerequisites
 
 When using {{site.prodname}} with a VM platform (e.g. OpenStack), obtaining IPv6
```
Always a comma before and after i.e. and e.g. In this case, since the e.g. is in parentheses, you just need one after it.

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+-  When it boots up, the VM should issue a DHCPv6 request for each of
+   its interfaces, so that it can learn the IPv6 addresses that
+   OpenStack has allocated for it.
+-  The VM must be configured to accept Router Advertisements.
```
"Router Advertisements" --> "router advertisements"

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+   its interfaces, so that it can learn the IPv6 addresses that
+   OpenStack has allocated for it.
+-  The VM must be configured to accept Router Advertisements.
+-  If it uses the widely deployed DHCP client from ISC, the VM must
```
You need to switch the it and the VM because the current sequence causes an unclear referent. "If it uses the widely deployed DHCP client from ISC, the VM " --> "If the VM uses the widely deployed DHCP client from ISC, it "

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,31 +1,158 @@
+   have a fix or workaround for [this known
+   issue](https://kb.isc.org/article/AA-01141/31/How-to-workaround-IPv6-prefix-length-issues-with-ISC-DHCP-clients.html).
 
 These requirements are not yet all met in common cloud images - but it
```
" - " --> "—"

i.e.: paste in actual em dash character and ensure that there are no spaces around it.

emanic commented on master/usage/ipv6.md:
```diff
@@ -49,35 +176,7 @@ changes will suffice to meet the requirements just listed.
-
-OpenStack Specific Details
---------------------------
+### Details of Enabling IPv6
```
This section should match previous...title and numbered list treatment

caseydavenport commented on master/usage/ipv6.md:
```diff
@@ -1,33 +1,147 @@
+- For kube-apiserver:
+  - The `--bind-address` or `--insecure-bind-address` flags should be set
+    to the appropriate IPv6 address or `::` for all IPv6 addresses on
+	the host.
```
formatting nit... but a lot of these end with a couple of words on a new line when I don't think they need to. Why not just have "the host" on the line above it?

caseydavenport commented on master/usage/ipv6.md:
```diff
@@ -1,33 +1,147 @@
+   deployment and save it as `calico.yaml`.
+1. If the ipam section in the `cni_network_config` in the `calico.yaml` file
+   has `"type": "calico-ipam"` then it should be modified to have
+   [IPv4 assigning disabled and IPv6 assigning
```
nit: probably better to say "should be modified to disable IPv4 assignments and enable IPv6 assignments."

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,33 +1,150 @@
+  the other hosts as well as an IPv4 address.
+- Each host must have the sysctl setting `net.ipv6.conf.all.forwarding`
+  setting it to `1`.  This ensures both Kubernetes service traffic
+  and Calico traffic is forwarded appropriately.
```
"Calico" --> "{{site.prodname}}"

emanic commented on master/usage/ipv6.md:
```diff
@@ -1,33 +1,150 @@
+   issue](https://kb.isc.org/article/AA-01141/31/How-to-workaround-IPv6-prefix-length-issues-with-ISC-DHCP-clients.html).
 
-These requirements are not yet all met in common cloud images - but it
+These requirements are not yet all met in common cloud images — but it
```
Still spaces around the dash

emanic commented on master/usage/ipv6.md:
```diff
@@ -49,35 +166,7 @@ changes will suffice to meet the requirements just listed.
-
-OpenStack Specific Details
---------------------------
+### Enabling IPv6 support in Calico
```
"Calico" --> "{{site.prodname}}"
